inference_lib/src/aqlm/inference_kernels/cuda_kernel.cpp

```diff
@@ -103,6 +132,124 @@ torch::Tensor code1x16_matmat(
   return output;
 }
 
+torch::Tensor code1x16_dequant(
+  const torch::Tensor& codes,
+  const torch::Tensor& codebooks,
+  const torch::Tensor& scales
+) {
+  check_use_bfloat16(codebooks);
+  auto in_features = codes.size(1) * 8;
+  auto out_features = scales.size(0);
+
+  auto weight = torch::zeros({out_features, in_features},
+    torch::TensorOptions()
+      .dtype(codebooks.dtype())
+      .device(codebooks.device())
+  );
+  code1x16_dequant_cuda(
+    codes.data_ptr(),
+    weight.data_ptr(),
+    codebooks.data_ptr(),
+    out_features,
+    in_features
+  );
+  weight *= scales.index({"...", 0, 0});
```

Comment:
scaling the weights is slower than scaling the output, since m is usually much smaller than k, at least in the server benchmarks I was doing.  For small m I saw a 30% difference for the 1x16.

Reply:
Sure! I forgot to change it when adapting dequant kernel into a matmul kernel.

Reply:
Are you sure that that would be numerically stable? From the looks of it, matmul results change by 2-10% on randomly sampled inputs and weights.

Reply:
Hmm, that's not good.

Reply:
But the generated texts looks identical to gemv... I'm not sure what's happening there

Reply:
I modified the code to scale the output instead